prov/util/src/util_mr_cache.c

```diff
@@ -123,6 +124,17 @@ void ofi_mr_cache_notify(struct ofi_mr_cache *cache, const void *addr, size_t le
 		assert(entry->cached);
 		util_mr_uncache_entry(cache, entry);
 
+		/* Due to merge region logic, it will identify adjacent
+		 * addresses as the same. We must ignore any adjacent
+		 * addresses, since they do not belong to the page fault.
+		 */
+		if (entry->use_cnt != 0 &&
+		    ((char *)entry->iov.iov_base +
+		    entry->iov.iov_len == addr ||
+		    (char *)addr + len == entry->iov.iov_base)) {
+			continue;
+		}
```

Comment:
If I understand correctly, this is a result of the behavior of the find() handler and only happens when merge_regions is enabled.  Correct?

I don't think we want the use_cnt check here, and there are a couple helper functions we can use (from ofi_iov.h).  I think this would work:
```
if (ofi_iov_left(&iov, &entry->iov) || ofi_iov_right(&iov, &entry->iov))
    continue;
```

Reply:
You are correct, the behavior of the find handler does this when merge_regions is enabled.

We do need the use_cnt check I believe because we are uncaching the region at this point. If use_cnt is 0, we can never find these registrations again I believe which will cause issues during clean-up. The code assumes that if it's uncached and use_cnt is 0, it's freed I believe. I was encountering many failures before I put in the use_cnt check. I think we can use helper functions though.

Reply:
The for() loop is finding all entries that contain the faulting region.  If the entry being returned is adjacent to the faulting region, it's a false hit.  I don't understand why we wouldn't want to just ignore those regions here.  They aren't related to the faulting region.  Your code ignores them when use_cnt is non-zero, but I don't see why that matters.

Reply:
I don't believe ofi_iov_left and ofi_iov_right can duplicate this behavior though...
ofi_iov_left compares iov1 base + len < iov2 base. It will return false if iov1 base + len = iov2 base.
We would want it to return true in this situation.

I don't believe ofi_iov_shifted_left or ofi_iov_shifted_right work here either.
I could just do ofi_iov_end(&iov) == entry->iov.iov_base || ofi_iov_end(&entry->iov) == iov.iov_base

Reply:
> The for() loop is finding all entries that contain the faulting region. If the entry being returned is adjacent to the faulting region, it's a false hit. I don't understand why we wouldn't want to just ignore those regions here. They aren't related to the faulting region. Your code ignores them when use_cnt is non-zero, but I don't see why that matters.

The issue in this situation is that the entry is being uncached (which it must do or else it will loop infinitely), with a use_cnt of 0, but not put on an lru list or freed. If we don't do that, we cannot find the entry again. Thus we can never free the entry.

Reply:
If the addr is the last byte of the entry (add == entry->iov.iov_base + entry->iov_len), then the entry *does* reference at least some portion of the faulting region.  In this case, we're only checking for a 1-byte overlap at either the start or end, which doesn't seem right.  I think we need a different set of logic here.

The check I gave above should be valid.  That will eliminate the false matches.

I believe the problem is that the cache merged regions without the application knowing about it.  So, the app may be accessing a merged region correctly (e.g. the upper half), even though a portion of the region (e.g. the lower half) is no longer valid.

Sigh... the simplest solution would be to remove the merge region feature.  If we want to keep the merge feature, then the assert(use_cnt == 0) is invalid.  The app may be using the portion of the region that is valid.  I don't think this is correct, but here's a first pass at handling this:

```
for (entry = ...; ; ...) {
    /* If the region is adjacent, it's a false match */
    if (ofi_iov_left(&iov, &entry->iov) || ofi_iov_right(&iov, &entry->iov))
        continue;

    /* following code is similar to removing an old entry after merging */
    util_mr_uncached_entry(cache, entry);

    if (entry->use_cnt == 0) {
        dlist_remove_init(...);
        util_mr_free_entry(...);
    }
}
```

Reply:
My math is wrong.  There's no overlap.

Reply:
Also for that loop, if you are using the same conditional as before, it would just loop infinitely since the continue would skip the uncache, it would find the same entry over and over again.

Reply:
If we can't skip false matches because of looping issues, we should probably drop the if iov statement completely then.  The faulting region could basically be anywhere inside of an entry that's actively in use.  The app just needs to be using the portion of the region that is still valid.  So, the assert() in the current code is wrong when merging is enabled.

Reply:
Actually, we don't even need merging for this problem to show up.

Reply:
The original logic was such that the merged region was a new and distinct memory registration, while the old region was placed onto a stale region list until the last reference dropped. Was that removed?

Reply:
That logic is there there.  I think the ultimate problem is that the assert(use_cnt == 0) is wrong.

Example, the app sends a 1 MB buffer.  The cache creates a new 1 MB region.  The app frees some portion of that buffer.  The app sends a 4 KB transfer that is part of the original 1 MB buffer.  A fault occurs on the portion that was freed.

Reply:
Now that I think about it, without merge regions, the search function is still not adequate.

For example, if registration X is from [0xA, 0xD] and registration Y is from [0xB,0xE], a fault event on [0xB,0xE] would detect registration Y but it would not detect registration X since the util_mr_find_within doesn't detect partial overlap, even if [0xB,0xD] is invalid.

Reply:
> That logic is there there. I think the ultimate problem is that the assert(use_cnt == 0) is wrong.
> 
> Example, the app sends a 1 MB buffer. The cache creates a new 1 MB region. The app frees some portion of that buffer. The app sends a 4 KB transfer that is part of the original 1 MB buffer. A fault occurs on the portion that was freed.

Sorry, I didn't follow this example. In this case, where is memory registration and deregistration happening?

If a libfabric provider is doing the registrations in-line, then it would look like this:

1. 1 MB sent by app
2. Provider registers a 1 MB buffer in the cache and sends
3. Provider deregisters 1 MB buffer and returns a send completion
4. Cache holds onto 1 MB registration but leaves use_cnt = 0
4. Portion of 1 MB buffer freed.
5. Userfault event occurs
6. Cache detects use_cnt = 0 on region and frees the entire 1 MB registration
7. Remaining portion (4 KB) sent
8. Provider registers a new region with the cache and sends
...

Reply:
The example would be:

1. 1 MB sent by app
2. Provider registers a 1 MB buffer in the cache and sends
3. Provider deregisters 1 MB buffer and returns a send completion
4. Cache holds onto 1 MB registration but leaves use_cnt = 0
5. App sends portion of buffer (4 KB) sent
6. Cache matches registered buffer, sets use_cnt = 1
7. Portion of 1 MB buffer freed.
8. Userfault event occurs

And you're right about the search function.  The rbtree sets the find function at init time.  We really want to have a specific search function here.  Maybe we can modify the rbtree implementation with a new search function that takes a new find callback as input?

Reply:
Okay, yeah that makes sense. That's pretty unfortunate because the assert(use_cnt==0) is actually useful in finding invalid behavior where a buffer is freed before memory is deregistered.

Yeah, I think we need to modify the rbtree find capabilities somehow.

Reply:
See #5074 for a proposal for handling the notifications.  It required updates to the rbtree to support a find function that overrides the stored compare handler.  I included a proposed fix for the actual issue discussed here.